### ISSUE TITLE
iOS - Heart Rate reading

### ISSUE THREAD

**Issue by luissantos95** [pangliang/miband-sdk-android#24]
HI,
I have been using your code to implement an iOS version of the app, so I can control the miBand.
Until now, I have been able to vibrate it and read the steps and battery.
I don't know why but I can't figure out how to read the heart rate. 
I tried to set the user info first but still doesn't work. I think it may be the set user code that is wrong. 
Can you help me write an iOS version to read the heart rate?

Thanks a lot.

**Comment by pangliang** [comment]
Bluetooth interface should be the same; as long as you know the device uuid, write BluetoothGattCharacteristic way are the same;

```java
byte[] data = userInfo.getBytes(device.getAddress());
this.io.writeCharacteristic(Profile.UUID_CHAR_USER_INFO, data, null);
```
In ios is the same, the UserInfo turn into a byte array, it is written to the uuid of the Characteristic `Profile.UUID_CHAR_USER_INFO` ("0000ff04-0000-1000-8000-00805f9b34fb")

I feel the need to pay attention may be big-endian or little-endian format; for example, on the model of UserInfo uid, is an int; may require little-endian flip; but I'm not sure

**Comment by luissantos95** [comment]
The only thing I think is wrong is the generation of the data from the user information. I wrote in swift but apparently I am not getting the same results as in Java
`private var uid: Int?
    private var gender: UInt8?
    private var age: UInt8?
    private var height: UInt8? // cm
    private var weight: UInt8?        // kg
    private var alias: String = ""
    private var type: UInt8?
    
    init(){}
    
    init(uid: Int,gender: Int, age: Int, height: Int, weight: Int, alias: String, type: Int) {
        self.uid = uid
        self.gender = UInt8(gender)
        self.age = UInt8(age)
        self.height = UInt8(height & 0xFF)
        self.weight = UInt8(weight)
        self.alias = alias
        self.type = UInt8(type)
    }
    
    func getBytes(mBTAddress: String) -> [UInt8] {
        var aliasBytes = [UInt8]();
        
        for char in self.alias.utf8{
            aliasBytes += [char]
        }
        var buffer = Array<UInt8>()
        
        buffer.append(UInt8((uid! & 0xff)))
        buffer.append(UInt8((uid! >> 8 & 0xff)))
        buffer.append(UInt8((uid! >> 16 & 0xff)))
        buffer.append(UInt8((uid! >> 24 & 0xff)))
        buffer.append(UInt8(self.gender!))
        buffer.append(UInt8(self.age!))
        buffer.append(UInt8(self.height!))
        buffer.append(UInt8(self.weight!))
        buffer.append(UInt8(self.type!))
        buffer.append(UInt8(4))
        buffer.append(UInt8(0))
        
        if (aliasBytes.count <= 8) {
            for byte in aliasBytes{
                buffer.append(byte)
            }
            for i in 0 ..< (8 - aliasBytes.count){
                buffer.append(UInt8(0))
                }
           
        } else {
            for i in 0..<8{
                buffer.append(aliasBytes[i])
            }
        }
        
        var crcSequence = Array<UInt8>(count: 19, repeatedValue: 0)
        for u in 0..<crcSequence.count{
            crcSequence[u] = buffer[u];
        }
        
        let myRange = Range<String.Index>(start: mBTAddress.startIndex.advancedBy(mBTAddress.characters.count - 2), end: mBTAddress.endIndex.advancedBy(-1))
        
        let crcb = UInt8(((getCRC8(crcSequence) ^ Int( mBTAddress.substringWithRange(myRange) )!) & 0xff))
        buffer.append(crcb)
        for _ in buffer.count..<20{
            buffer.append(UInt8())
        }
        return buffer
    }
    
    private func getCRC8(seq: [UInt8]) -> Int {
        var len = seq.count
        var i = 0;
        var crc: UInt8 = 0x00;
        
        while (len-- > 0) {
            var extract = UInt8(seq[i++]);
            for _ in 0..<8 {
                var sum = UInt8((UInt8(crc) ^ UInt8(extract)));
                sum = UInt8(((sum & 0xff) & 0x01));
                crc = UInt8((UInt8(crc & 0xff) >> 1) & UInt8(0b0111111))
                if (sum != 0) {
                   crc = UInt8((crc & 0xff) ^ 0x8c);
                }
                extract = (UInt8(extract & 0xff) & 0b0111111);
            }
        }
        return Int((crc & 0xff));
    }`

**Comment by pangliang** [comment]
try int8, JAVA have not unsigned number.  byte -> int8;  short -> int16  i think.
Use uint8 may be a problem

**Comment by luissantos95** [comment]
OK, so after more than 20 hours spent on this I figured out how to manage all the logic with swift.
Now, when I set the user info (with type = 1), the band vibrates and the LED start blinking (the steps are reset to zero). Even after that, I can not start the heart rate reading (writing START_HEART_RATE_SCAN). I even tried using type 0, but when I do, nothing happened (LED did not blink or steps where reset)

**Comment by pangliang** [comment]
Have you noticed, writeCharacteristic there are two arguments, serviceUUID and CharacteristicUUID, HEARTRATE **serviceUUID** is the special one; 
```java
MiBand.this.io.writeCharacteristic (
Profile.UUID_SERVICE_HEARTRATE,   // Unlike other functions
Profile.UUID_CHAR_HEARTRATE, 
Protocol.START_HEART_RATE_SCAN, null);
```

**Comment by luissantos95** [comment]
Yes I did. But it make no difference when writing with the Core Bluetooth Library for iOS.
As I said, I can write the user info (when type is 0). The LEDs blink and the band vibrates. After that, it starts the connection all over again, reseting the number of steps. I tried to initialize the heart rate scan after that but I couldn't. I even tried to write the user info before (with type 1), but it did not work.

**Comment by pangliang** [comment]
the band vibrates and the LED start blinking, this time did you tapped the bracelet?
setNormalNotifyListener() and when setUserInfo successful, you can be notified of a value of (int)3.

**Comment by luissantos95** [comment]
I tapped the band and all LED's blinked together one time. 
When I do setUserInfo (type 1), I get a notify of 10 (Int). When I set with type 0, I get 5 (Int).

**Comment by luissantos95** [comment]
Maybe I found the problem.
I have a M1S and my Band firmware is 4.16.3.7 and the Heart sensor firmware is 1.3.76.20.
Maybe they changed the bytes that have to be sent to start the heart sensor scan

**Comment by luissantos95** [comment]
Got it.... To start reading, send: 15,2,1 instead of 21,2,1 :)

**Comment by AlexeyPelekh** [comment]
@luissantos95 is it possible to get some of your sources?

**Comment by TSG9876** [comment]
Hi , just the confirm , did you call 15,2,1 after heartscan or before , I call this after heartscan to return heartrate data .

I am now trying to determine how to know  whether the Mi Band heartscan (green led blinking) is busy scanning or not?

thanks
